src/Symfony/Component/HttpKernel/HttpKernel.php

```diff
@@ -36,19 +38,29 @@ class HttpKernel implements HttpKernelInterface, TerminableInterface
     protected $dispatcher;
     protected $resolver;
     protected $requestStack;
+    private $argumentResolver;
 
     /**
      * Constructor.
      *
-     * @param EventDispatcherInterface    $dispatcher   An EventDispatcherInterface instance
-     * @param ControllerResolverInterface $resolver     A ControllerResolverInterface instance
-     * @param RequestStack                $requestStack A stack for master/sub requests
+     * @param EventDispatcherInterface    $dispatcher       An EventDispatcherInterface instance
+     * @param ControllerResolverInterface $resolver         A ControllerResolverInterface instance
+     * @param RequestStack                $requestStack     A stack for master/sub requests
+     * @param ArgumentResolverInterface   $argumentResolver An ArgumentResolverInterface instance
      */
-    public function __construct(EventDispatcherInterface $dispatcher, ControllerResolverInterface $resolver, RequestStack $requestStack = null)
+    public function __construct(EventDispatcherInterface $dispatcher, ControllerResolverInterface $resolver, RequestStack $requestStack = null, ArgumentResolverInterface $argumentResolver = null)
     {
         $this->dispatcher = $dispatcher;
         $this->resolver = $resolver;
         $this->requestStack = $requestStack ?: new RequestStack();
+
+        if (null === $argumentResolver) {
+            // fallback in case of deprecations
+            $argumentResolver = $resolver instanceof ArgumentResolverInterface ? $resolver : new ArgumentResolver();
```

Comment:
I would replace this by a bc layer more complex as what was done in 2.x but allowing to keep the current interface:
```php
if (ControllerResolver::class !== get_class($resolver)
            && ($reflector = new \ReflectionMethod($resolver, 'getArguments'))
            && ControllerResolver::class !== $reflector->class
            && ($reflector = new \ReflectionMethod($resolver, 'doGetArguments'))
            && ControllerResolver::class !== $reflector->class
        ) {
    @trigger_error('...');
    $argumentResolver = $resolver;
} else {
    $argumentResolver = new ArgumentResolver();
}
```

What do you think ?